### PR TITLE
WIP (fixing some bugs): Add Teleportation gun/grenade/rifle

### PR DESCRIPTION
**Pull request by Ryozuki**
This PR is like the second version of https://github.com/ddnet/ddnet/pull/677 but better of course.

Features:
- Teleport gun: Teleports you where it collided (tees included, in a special way). Doesn't teleport if the collided tile is an untele tile.
- Teleport grenade: Teleports you where it collided (tees included, in a special way). Doesn't teleport if the collided tile is an untele tile.
- Teleport rifle (laser): Teleports you where the laser runs out of energy or if it collides with a tee. Doesn't teleport if the laser bounced in an untele tile.

I think this adds a lot of game-play, it also may need a lot of testing to check if there are bugs. I think the best way to test this is that mappers do some maps. So here I leave some compiled binaries:
- [ddnet-tele-windows.zip](https://github.com/ddnet/ddnet/files/1798492/ddnet-tele-windows.zip)

*Note: untele tiles should be placed inside unhookable/hookable.*

Credit to Ravie, forsaken and hi_leute_gll for the images, and timakro for helping.

### CODE REVIEW

**Comment by heinrich5991** [comment]
I'm still against this. :/

The teleport gun, grenade, rifle cannot be predicted client-side, thus it will have laggy gameplay. This goes together with very abrupt gameplay because it's not even always clear when the bullets/laser will finally do the teleport.

**Comment by Ryozuki** [comment]
Oh, it's this what you mean when you said `Grandpa is against a merge of new gameplay features, in general.` ?

**Comment by heinrich5991** [comment]
No, I haven't tested it yet, I guess I should.

**Comment by heinrich5991** [comment]
I joined the test server. Laser teleport seemed to be bugged a little, when falling, shooting up or down always teleported me.

**Comment by Ryozuki** [comment]
@heinrich5991 it's not bugged, if you would read what my inital comment says... god.

If laser bounces atleast **once** on a anti-tele tile it wont teleport you. And I'm sure that's what happened to you.

**Comment by heinrich5991** [comment]
Another thing why I see as problematic to merge: We don't have a process for merging new gameplay features. This is a gameplay feature with a non-obvious implementation, especially about where you place the tee when it the teleport thing hits a wall, where you place the tee when the teleport thing hits another tee. This is different from e.g. merging another kind of teleporter where the implementation is pretty obvious: move the character/projectile/whatever to the location; but even for this, there are some interesting variables, like: does the character/projectile/whatever keep its speed/direction?

So far people and ourselves expected that you're not allowed to change gameplay features once they're used in maps, and some people even go as far as saying that [bugs in existing features cannot be fixed](https://forum.ddnet.tw/viewtopic.php?f=68&t=5914#p59374). That means by merging this, some people would expect the feature to be frozen forever, with all its bugs and design choices. This is a bad thing for a pretty new feature IMO, most of the previously added features to DDNet were pretty obvious in the sense that there is likely only "one" correct implementation for them, e.g. compare checkpoint teleporters.

I don't know a good solution to this, as I clearly see that you want this feature in DDNet. Perhaps it would be best to experiment with this outside of mainline DDNet?

**Comment by gamertyp** [comment]
That's one of the reasons, why I suggested Ryozuki to make a test phase - which is what he did. So far I don't know about anyone complaining that the implementation is "wrong" in the sense of a wrong interpretation of how the feature should work. Also, the circumstances that people can have different opinions shouldn't be a reason to decline more complex features in general. This really sounds like:

> Grandpa is against a merge of new gameplay features, in general

PS: There is only one bug I know of, which is caused by the feature. I already told Ryozuki.

**Comment by heinrich5991** [comment]
>This really sounds like:
>>Grandpa is against a merge of new gameplay features, in general

Yes. I went and thought about *why* that is the case, and I found the reason and laid it out in my post.

**Comment by def-** [comment]
I'd like to merge this and we can release a map with it. Then after a few weeks check how well it was received.

File: src/game/server/entities/laser.cpp [hunk 1]
```diff
@@ -164,6 +165,52 @@ void CLaser::DoBounce()
...
+			int64_t TeamMask = pOwnerChar->Teams()->TeamMask(pOwnerChar->Team(), -1, m_Owner);
+
+			vec2 PossiblePos;
+			bool Found = false;
+
+			bool pDontHitSelf = g_Config.m_SvOldLaser || (m_Bounces == 0 && !m_WasTele);
+			vec2 At;
+			CCharacter *pHit;
+			if(pOwnerChar ? (!(pOwnerChar->m_Hit&CCharacter::DISABLE_HIT_RIFLE) && m_Type == WEAPON_RIFLE) : g_Config.m_SvHit)
+				pHit = GameServer()->m_World.IntersectCharacter(m_Pos, To, 0.f, At, pDontHitSelf ? pOwnerChar : 0, m_Owner);
```
Comment (Learath2) [inline src/game/server/entities/laser.cpp]: Braces

File: src/game/server/entities/laser.cpp [hunk 2]
```diff
@@ -164,6 +165,52 @@ void CLaser::DoBounce()
...
+		{
+			int64_t TeamMask = pOwnerChar->Teams()->TeamMask(pOwnerChar->Team(), -1, m_Owner);
+
+			vec2 PossiblePos;
+			bool Found = false;
+
+			bool pDontHitSelf = g_Config.m_SvOldLaser || (m_Bounces == 0 && !m_WasTele);
+			vec2 At;
+			CCharacter *pHit;
+			if(pOwnerChar ? (!(pOwnerChar->m_Hit&CCharacter::DISABLE_HIT_RIFLE) && m_Type == WEAPON_RIFLE) : g_Config.m_SvHit)
```
Comment (Learath2) [inline src/game/server/entities/laser.cpp]: Weird to read

File: src/game/server/entities/projectile.cpp [hunk 3]
```diff
@@ -160,6 +159,61 @@ void CProjectile::Tick()
 		}
 		else if(pTargetChr && m_Freeze && ((m_Layer == LAYER_SWITCH && GameServer()->Collision()->m_pSwitchers[m_Number].m_Status[pTargetChr->Team()]) || m_Layer != LAYER_SWITCH))
 			pTargetChr->Freeze();
+
+		if (pOwnerChar && ColPos && !GameLayerClipped(ColPos))
+		{
+			bool IsTeleValid = false;
+			switch (m_Type)
+			{
+				case WEAPON_GRENADE:
```
Comment (Learath2) [inline src/game/server/entities/projectile.cpp]: Wrong indentation, unnecessary block.

**Comment by Ryozuki** [comment]
Give, me some time, i'll fix the bugs hileute found, the style and anything else i find in the meantime. I'll tell here when it's ready to merge.

**Comment by timakro** [comment]
I tried it and liked it. It's really a matter of how mappers will be making use of the feature. Client side prediction would be possible, right? If people enjoy this that could be implemented. I attached the map in case you'd like to give it a go as well (for some insane reason GitHub only allows certain file types, it's a map, just remove the `.txt` extension).
[TeleTest.map.txt](https://github.com/ddnet/ddnet/files/1872956/TeleTest.map.txt)

**Comment by Ryozuki** [comment]
Yeah, i would implement prediction after knowing if community likes it. It isn't that bad without predection right now. I know of a guy who finished the map attached with 300 ping.

**Comment by Learath2** [comment]
Is this still WIP?

**Comment by Ryozuki** [comment]
yes, and u should remove ur changes request as its fixed and i already told you time ago in discord.

**Comment by Learath2** [comment]
You are still missing braces unles I'm missing sth.

**Comment by Ryozuki** [comment]
https://github.com/ddnet/ddnet/blob/master/src/game/collision.cpp#L95

I'm not gonna add braces for single line ifs

**Comment by Ryozuki** [comment]
Closing as ddnet right now doesn't seem to like others work, since they think giving credit is trolling.

**Pull request closed without merging**